openedx/openedx-platform
# fix: refactor MakoService to allow specifying template more explicitly (Take 2)

## Description

This is a revised version of #33061, which was reverted by #33070.

See the original PR for details.

This PR fixes:
* the "Manage Access" menu in Studio is not working consistently
* the problematic ambiguity in the MakoService's `render_template` design that has been causing bugs like ^
* the Instructor Dashboard was not loading with the previous fix in place

**What's different in this PR?**

This PR is exactly the same, but includes eb9b13a00a0757456c9cfe5149d092b718b8b448 to fix the instructor dashboard issue.

## Error details

**What was the instructor dashboard issue?**

*If* you had the "bulk email" tab of the instructor dashboard enabled, then anytime you accessed the instructor dashboard (regardless of which tab), you would trigger [an old hack that renders a mock `HtmlBlock`'s `studio_view` in the LMS to act as an HTML editor](https://github.com/openedx/edx-platform/blob/a4b11c02e0171b977abbd0c26ea7fae7be7ed36c/lms/djangoapps/instructor/views/instructor_dashboard.py#L667) for composing the email.

Because I assumed that nobody would be using `studio_view` in the LMS (which they shouldn't, to be clear!), my original PR would throw an error: "Cannot access CMS templates from the LMS" when the fake HtmlBlock's `studio_view` tried to render its HTML editor template, [`html-edit.html`](https://github.com/openedx/edx-platform/blob/master/cms/templates/widgets/html-edit.html).

Why did this work before my change? Well, the behavior before was ambiguous. Normally, when the HTML block editor was run in Studio, it would assume that the CMS template `cms/templates/widgets/html-edit.html` was wanted, and use that. But in this one case, the bulk email page was actually rendering `studio_view` from the LMS, which was then using a similar but different template, `lms/templates/widgets/html-edit.html`. Oddly, even when this second template was first added ten years ago (3ea2b24b988016d875d3ebcba5fd319c61db0640), it was a bit different from the CMS version and has been slightly different ever since, used only by this bulk email page as far as I know.

Another contributing factor is that the code of the current instructor dashboard is not properly isolated with error boundaries, so a template error on the Bulk Email tab causes the whole instructor dashboard to become unavailable. ☹️

**How did I fix it?**

For now I only removed one layer of hacks from this pile, by not actually invoking `HtmlBlock` nor the XBlock runtime at all. Instead, the Bulk Email page now just directly renders the templates that it needs to make the HTML editor work. I believe the plan is to replace the whole Instructor Dashboard with a pluggable, properly isolated MFE version (right?), so I don't think it would make sense to spend the time to improve this server-rendered template version any further.

## Testing instructions

1. Check out this branch but excluding the final commit (i.e. exclude the commit that fixes the instructor dashboard issue)
2. Go to your LMS Django admin "Bulk email flags" section (e.g. http://local.overhang.io:8000/admin/bulk_email/bulkemailflag/ ) and add a new configuration that is "Enabled: Yes" and "Require course email auth: No"
3. Go to any course's instructor dashboard in the LMS. Observe the error.
4. Check out this branch including the new fix commit.
5. Go back to the instructor dashboard. Observe that it's working. Compose a bulk email including rich text and use the tool to send it to "Myself"
6. Go to "Course emails" in the Django admin (/admin/bulk_email/courseemail/) and view the sent email. Check that the HTML and plain text versions are both correct, indicating that the rich text editor is working well.

## Deadline

Sooner is better as this fixes the "Manage Access" CAT-2 issue.

## Other information

Private ref MNG-3884
